dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java

```diff
@@ -446,32 +449,41 @@ public TypedFormula<BooleanType, BooleanFormula> visitIntCmpExpression(IntCmpExp
         @Override
         public TypedFormula<IntegerType, ?> visitIntConcat(IntConcat expr) {
             Preconditions.checkArgument(!expr.getOperands().isEmpty());
-
+            final List<? extends TypedFormula<IntegerType, ?>> operands = expr.getOperands().stream()
+                    .map(this::encodeIntegerExpr)
+                    .toList();
+            Formula enc = operands.get(0).formula();
             if (context.useIntegers) {
-                throw new UnsupportedOperationException("Cannot encode bitwise concatenation on mathematical integers.");
+                final IntegerFormulaManager imgr = fmgr.getIntegerFormulaManager();
+                long offset = types.getMemorySizeInBits(operands.get(0).type());
+                for (TypedFormula<IntegerType, ?> op : operands.subList(1, operands.size())) {
+                    final IntegerFormula offsetValue = imgr.makeNumber(1L << offset);
```

Comment:
You should do the shift on `BigInteger`, otherwise the code may fail on larger concatenations/concatenation of larger integers.